lib/connections/svc.go

```diff
@@ -165,16 +177,16 @@ next:
 		// clients between the same NAT gateway, and global discovery.
 		if remoteID == s.myID {
 			l.Infof("Connected to myself (%s) - should not happen", remoteID)
-			c.Conn.Close()
+			c.Close()
 			continue
 		}
 
-		// If we have a relay connection, and the new incoming connection is
-		// not a relay connection, we should drop that, and prefer the this one.
+		// If we have a connection, and the new incoming connection is
+		// of a higher priority, we should drop that, and prefer the this one.
 		s.mut.RLock()
-		ct, ok := s.connType[remoteID]
+		ct, ok := s.currentConnection[remoteID]
 		s.mut.RUnlock()
-		if ok && !ct.IsDirect() && c.Type.IsDirect() {
+		if ok && ct.Priority < c.Priority {
```

Comment:
Possibly, but https://github.com/syncthing/syncthing/blob/master/lib/discover/cache.go#L282 and things like nice levels, network priority marking etc work in the other direction (lower = better).